### PR TITLE
fix(jenkins-status): checks for Jenkins status before promoting

### PR DESCRIPTION
**Pull request by kishansagathiya**
Fix for https://github.com/openshiftio/openshift.io/issues/3356

### CODE REVIEW

**Comment by alien-ike** [comment]
### Ike Plugins (test-keeper)

Thank you @kishansagathiya for this contribution!



It appears that no tests have been added or updated in this PR.

Automated tests give us confidence in shipping reliable software. Please add some as part of this change.

If you are an admin or the reviewer of this PR and you are sure that no test is needed then you can use the command `/ok-without-tests` as a comment to make the status green.


For more information please head over to official [documentation](http://arquillian.org/ike-prow-plugins/#_test_keeper_plugin). You can find there how to configure the plugin - for example exclude certain file types so if PR contains only them it won't be checked.

File: src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts [hunk 1]
```diff
@@ -84,13 +86,35 @@ export class ApplicationsListItemComponent implements DoCheck, OnInit {
     }
     let inputAction = input.build.firstPendingInputAction;
     if (isValidInputAction(inputAction) && input.build.jenkinsNamespace) {
+      console.log('####-12');
```
Comment (kishansagathiya) [inline src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: To be removed

File: src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts [hunk 2]
```diff
@@ -84,13 +86,35 @@ export class ApplicationsListItemComponent implements DoCheck, OnInit {
...
+      this.checkIfJenkinsIsUp();
       this.inputActionDialog.build = input.build;
       this.inputActionDialog.inputAction = inputAction;
       this.inputActionDialog.stage = input.stage;
-      this.inputActionDialog.proceed();
     }
   }
 
+  checkIfJenkinsIsUp(): any {
+  let url = 'https://jenkins.openshift.io';
```
Comment (kishansagathiya) [inline src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: Not sure how to retrieve this URL instead of hardcoding it

**Comment by joshuawilson** [comment]
Why is the package-lock file being updated?

**Comment by kishansagathiya** [comment]
Do not review yet. I just realized that promote button is at two different places. I have to change the code.

File: src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts [hunk 3]
```diff
@@ -84,13 +86,35 @@ export class ApplicationsListItemComponent implements DoCheck, OnInit {
...
+  this.http.get(url)
+    .subscribe((response) => {
+      if (response.status == 200) {
+        console.log('Got ' + response.status + ', Jenkins is up and running...');
+        this.inputActionDialog.proceed();
+        return true;
+      }
+      if (response.status == 307 || response.status == 302) {
+        console.log('Got ' + response.status + ', Jenkins Proxy is logging you in...');
```
Comment (gastaldi) [inline src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: No indication in the UI that this is happening?

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 4]
```diff
@@ -84,4 +84,23 @@ export class InputActionDialog {
...
+  checkIfJenkinsIsUp() {
+    let url = 'https://jenkins.openshift.io';
+    this.http.get(url)
+      .subscribe((response) => {
+        if (response.status == 200) {
+          console.log('Got ' + response.status + ', Jenkins is up and running...');
+          this.invokeUrl(this.inputAction.proceedUrl);
+          return;
+        } else if (response.status == 307 || response.status == 302) {
+          console.log('Got ' + response.status + ', Jenkins Proxy is logging you in...');
```
Comment (gastaldi) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: No indication in the UI that this happened?

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 5]
```diff
@@ -84,4 +84,23 @@ export class InputActionDialog {
   close() {
     this.modal.close();
   }
+
+  checkIfJenkinsIsUp() {
+    let url = 'https://jenkins.openshift.io';
```
Comment (gastaldi) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: A new environment needs to be added to F8UI (the Jenkins URL)

Comment (kishansagathiya) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: I wanted to inject an environment variable like https://github.com/fabric8-ui/fabric8-ui/blob/master/src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts#L5-L26
Where and how do I do this

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 4 above)
Comment (rohitkrai03) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: @gastaldi We could probably use NotificationService to show these messages to user rather than just logging it to console. @kishansagathiya What do you think?

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 5 above)
Comment (gastaldi) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: @kishansagathiya I think @edewit or @joshuawilson may know this better than me :wink:

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 6]
```diff
@@ -84,4 +84,25 @@ export class InputActionDialog {
...
+        } else if (response.status == 307 || response.status == 302) {
+          console.log('Got ' + response.status + ', Jenkins Proxy is logging you in...');
+        } else if (response.status == 503) {
+          console.log('Got ' + response.status + ', Cluster resources capacity is full. waiting along..');
+        } else if (response.status == 202) {
+          console.log('Got 202, Jenkins is currently idled. Please wait while we start it. waiting along..');
+        } else {
+          console.error ('Got status ' + response.status);
+        }
+        setTimeout (this.checkIfJenkinsIsUp(), 10000);
```
Comment (lordofthejars) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: Looks like infinite loop isn't it? It just finishes if we get a `200`. Shouldn't we finish this loop if we get an unexpected error? And also add some feedback to user? Something like spinning animation in case of unidling?

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 4 above)
Comment (lordofthejars) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: yes some kind of feedback will be great so user knows that he has pushed the button and we are working on it.

**Comment by kishansagathiya** [comment]
@joshuawilson 
Have tested this locally, 
With prod-preview Jenkins proxy I was able to get the proper status code
but on prod Jenkins proxy it gave me Cross-Origin Resource Sharing error.
Since once deployed domain would be same, so we won't see this. There might be few things that I might be missing in this PR (like hardcoded URL). I don't know how to do that.
Somebody with UI experience would be able to do it much better and faster than.

File: src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts [hunk 7]
```diff
@@ -90,7 +90,6 @@ export class ApplicationsListItemComponent implements DoCheck, OnInit {
       this.inputActionDialog.proceed();
     }
   }
-
```
Comment (joshuawilson) [inline src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: why was this line removed? It is there for readability?

Comment (kishansagathiya) [reply src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: Unintentional. I had changed this file. Might have left this. Let get it how it was

Comment (kishansagathiya) [reply src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: Yup, I know why. It gives me linting errors without that.
93,1 trailing whitespace

**Comment by jiekang** [comment]
In theory CORS won't be hit on production since it will be openshift.io making a request to jenkins.openshift.io. Unfortunately allowing CORS is a server-side thing so testing it locally does not seem possible.

**Comment by rohitkrai03** [comment]
@jiekang Since most of the API URLs we have come from external dependencies like ngx-login-client or ngx-fabric8-wit, i am not sure what's the best place to put the jenkins URL.

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 4 above)
Comment (maxandersen) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: lets not reveal internal implementation details like "Jenkins proxy" to user in the UI here. 

Better to just say "Connecting to Jenkins..." or similar.

**Comment by jiekang** [comment]
Hmm, a few issues varying in degree of severity. @rohitkrai03 

1. Basic GET request to jenkins.openshift.io **always** results in 307 response, regardless of state of jenkins pods. I don't know there's another endpoint that can be used but this one may not be correct. The request also doesn't have any auth header but I'm doubtful it would change anything.
2. Clicking promote no longer runs `this.close()` so the modal is not closed. It should close at the end, regardless of success.
3. I agree with your earlier comment regarding using NotificationService to inform users what happened after pressing promote.
4. This is a band-aid and I hope the issue is tracked for a proper long-term solution. In the event a user closes the tab containing this application, this repeated setTimeout will be destroyed and Promote will not occur.

**Comment by jiekang** [comment]
Also when I ran locally targeting production, I was able to see the request made when hitting 'Promote' and I didn't have a CORS issue, so I guess my earlier comment re. CORS can be disregarded. Sorry about that;;

**Comment by kishansagathiya** [comment]
jenkins.openshift.io?token_json=$OSIO_TOKEN_JSON should return status code
other than 307

**Comment by jiekang** [comment]
Ah it looks like initial request is okay but the CORS comes in due to the 307 as it doesn't allow the redirect to `https://auth.openshift.io/api/login?redirect=https:%2F%2Fjenkins.openshift.io%2F`

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 8]
```diff
@@ -84,4 +86,42 @@ export class InputActionDialog {
...
+            type: NotificationType.SUCCESS
+          } as Notification);
+          this.invokeUrl(this.inputAction.proceedUrl);
+          return;
+        } else if (response.status == 307 || response.status == 302) {
+          this.notifications.message({
+            message: `Got ${response.status}, Connecting to Jenkins...`,
+            type: NotificationType.INFO
+          } as Notification);
+        } else if (response.status == 503) {
```
Comment (kishansagathiya) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: We need one more retry logic for status is 504 Gateway Timeout

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-9 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

**Comment by kishansagathiya** [comment]
Todo
1. Add one more condition for retry (if status code is 504 Gateway timeout)
2. Remove hardcoded url and  inject it from environment variable 
3 . Add Tests

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-10 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-11 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 9]
```diff
@@ -84,4 +86,55 @@ export class InputActionDialog {
...
+            };
+            break;
+          default:
+            notification = {
+              message: `Got status ${response.status}`,
+              type: NotificationType.DANGER
+            };
+            break;
+        }
+        this.notifications.message(notification as Notification);
```
Comment (lordofthejars) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: Notice that now there is no return as it happens in if/else so infinite loop might occurs because break only occurs in switch context.

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-12 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-13 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 10]
```diff
@@ -88,7 +88,8 @@ export class InputActionDialog {
   }
 
   checkIfJenkinsIsUpAndProceedURL() {
-    let url = 'https://jenkins.openshift.io';
+    let url = 'https://jenkins.prod-preview.openshift.io';
```
[the remark this answers is not in the document]
Comment (joshuawilson) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: We need to set the URL via the env var that are set with SaaSHearder in https://github.com/openshiftio/saas-openshiftio/blob/master/dsaas-services/fabric8-ui.yaml#L10
and referenced like https://github.com/fabric8-ui/fabric8-ui/blob/master/openshift/fabric8-ui.app.yaml#L57

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-15 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: Looked everywhere but didn't find where we are setting the jenkins proxy url.
What we found is that `this.build.jenkinsBuildURL` in the current context has the right jenkins proxy url which can be hit to un-idle.


```
let url = this.build.jenkinsBuildURL + '?token_json=' + JSON.stringify(token_json);
``` 
is being used to call the jenkins proxy with the user's token - this will trigger un-idling.

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 9 above)
Comment (joshuawilson) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: This only sets the `notification` variable. There is nothing returned at all. The `default` sets it too. This looks correct to me.

File: src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts (hunk 7 above)
Comment (joshuawilson) [reply src/app/dashboard-widgets/applications-widget/applications-list-item/applications-list-item.component.ts]: Is there a single white space in the line? If so, just remove the single space and leave the line.

**Comment by kishansagathiya** [comment]
@joshuawilson 
Let's **NOT** merge this until the demo is done. This should not affect the demo. This might affect users during workshops and new users which will be added.

**Comment by rohitkrai03** [comment]
@kishansagathiya I've added a few tests for the component. I think all the todo items have been taken care of now. Do let me know if there is anything else that needs to be done. cc: @joshuawilson

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-16 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 11]
```diff
@@ -22,6 +25,7 @@ export class InputActionDialog {
   @ViewChild('inputModal') modal: any;
 
   constructor(private http: Http,
+              private notifications: Notifications,
               private authService: AuthenticationService,
               @Inject(FABRIC8_FORGE_API_URL) private forgeApiUrl: string
```
Comment (gastaldi) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: I believe `FABRIC8_FORGE_API_URL` needs to be completely removed, since it is no longer used

**Comment by joshuawilson** [comment]
@jiekang can you please take another look too.

**Comment by jiekang** [comment]
Okay, taking a look now.

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 12]
```diff
@@ -84,4 +88,76 @@ export class InputActionDialog {
...
+            break;
+          case 504:
+            notification = {
+              message: `Got ${response.status}, Request timed out. Trying again.`,
+              type: NotificationType.WARNING
+            };
+            break;
+          case 202:
+            notification = {
+              message: `Got ${response.status}, Jenkins is currently idled. Please wait while we start it. waiting along.`,
```
Comment (jiekang) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: s/Please wait while we start it. waiting along./Please wait while it starts./

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 13]
```diff
@@ -84,4 +88,76 @@ export class InputActionDialog {
...
+    };
+    let url = this.build.jenkinsBuildURL + '?token_json=' + encodeURIComponent(JSON.stringify(token_json));
+    return this.http
+      .get(url)
+      .map((response: Response) => {
+        return response;
+      })
+      .catch((error) => {
+        this.notifications.message({
+          message: 'There seems to be a problem with Jenkins',
```
Comment (jiekang) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: In a local instance connected to production, a redirect still occurs that results in this error being hit and unidling behavior not occurring. 

pipelines-redirect

The only way to fix this is to configure the server to send the right response header, but that's not an option. It may work in production though.

It looks like the jenkins proxy is configured s.t.
1. If a request to jenkins.openshift.io/*, redirect to auth.
2. If redirected to auth, redirect back to jenkins.openshift.io/* with ?token set
3. If a request to jenkins.openshift.io/* with ?token set, redirect to jenkins.openshift.io/*

It would be nice if there was an endpoint that didn't involve redirecting as long as the auth header was set with the correct token. Maybe the entire redirect mechanism can be looked at too, it's a bit long and browsers will show 'Too many redirects' even though it eventually goes to the right place. I digress...

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 14]
```diff
@@ -84,4 +88,76 @@ export class InputActionDialog {
...
+
+  callJenkins(): Observable<Response> {
+    const token_json = {
+      access_token: this.authService.getToken(),
+      token_type: 'Bearer'
+    };
+    let url = this.build.jenkinsBuildURL + '?token_json=' + encodeURIComponent(JSON.stringify(token_json));
+    return this.http
+      .get(url)
+      .map((response: Response) => {
```
Comment (jiekang) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: Using `map` to take `response` and return` response` is redundant, `http.get()` already returns the Response object, no?

`http.get(...).error(...)` should do the same thing.

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts (hunk 13 above)
Comment (jiekang) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: Also something like: s/There seems to be a problem with Jenkins/There is a problem accessing Jenkins./

It's best to avoid 'probably' messages ('seems' in this case) for the user. It's not as polished.

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 15]
```diff
@@ -84,4 +88,76 @@ export class InputActionDialog {
...
+          case 200:
+            notification = {
+              message: `Got ${response.status}, Jenkins is up and running.`,
+              type: NotificationType.SUCCESS
+            };
+            this.notifications.message(notification as Notification);
+            this.invokeUrl(this.inputAction.proceedUrl);
+            return;
+          case 307:
+          case 302:
```
Comment (jiekang) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: I think the default behavior of the http library when hitting a 302 from `get(...)` is to follow the redirect automatically. I think we've observed this in practice here as well. So theoretically, it won't ever be 302. Up to you to keep this or not.

**Comment by rohitkrai03** [comment]
@jiekang I tried writing tests for other status codes but ran into a problem because all the others cases call `setTimeout(this.checkIfJenkinsIsUpAndProceedURL(), 10000);` which makes it run out of stack in a test case. Is there a way to test this?

**Comment by jiekang** [comment]
@rohitkrai03 Oh that's fun :) I think that might also be an issue in production. If jenkins never hits 200, will a user also run out of stack, not just in tests? Maybe it would be best to only retry a set number of times, maybe 5 or 10 and then prompt the user to try again later. Thanks for trying to write tests here!

File: src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts [hunk 16]
```diff
@@ -84,4 +88,73 @@ export class InputActionDialog {
...
+              type: NotificationType.WARNING
+            };
+            break;
+          default:
+            notification = {
+              message: `Got status ${response.status}`,
+              type: NotificationType.DANGER
+            };
+            this.notifications.message(notification as Notification);
+            this.close();
```
Comment (jiekang) [inline src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: This is within an asynchronous block called from `checkIfJenkinsIsUpAndProceedURL()`. This function always calls `this.close()` before returning (line 141) so I don't think the default case needs to also call `this.close()`. Thoughts?

Comment (rohitkrai03) [reply src/a-runtime-console/kubernetes/ui/pipeline/input-action-dialog/input-action-dialog.component.ts]: @jiekang I agree. Removing `this.close()` from default case. I added it earlier thinking that if default case returns then the modal might not be closed but that's not the case since switch is under .subscribe() and `this.close()` in the end always gets called.

**Comment by jiekang** [comment]
@rohitkrai03 yes I think limiting retries would be much preferred then. If you want to dive into Observables, I think you can restructure this into an Observable that retries via `observable.retryWhen(...)` instead of using a setTimeout on the function. That setTimeout will also result in `this.close()` being called many times, which is probably safe but not the cleanest.

Also now that I think about it, the behavior of Observable returned by `http.get()` might need to be investigated. If it recieves an error status 4xx or 5xx, will it call the success function or the error function? I have a feeling it might call the error function so the `switch` for 4xx and 5xx errors is never hit.

**Comment by fabric8cd** [comment]
@kishansagathiya fabric8/fabric8-ui:SNAPSHOT-PR-2889-18 fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2889-fabric8-ui.badger.fabric8.io

**Comment by kishansagathiya** [comment]
Does https://github.com/fabric8-services/fabric8-jenkins-proxy/issues/257 affect merge of this PR in any way?

**Comment by jiekang** [comment]
I think if https://github.com/fabric8-services/fabric8-jenkins-proxy/issues/257 can be resolved positively than it's likely that this merge will actually work in production or prod-preview. There's still no way to test this locally but it's not necessarily a blocker as merging and testing on prod-preview is possible.

**Comment by joshuawilson** [comment]
I'm going to merge this since we need to test it in prod-preview to fully check it. @rohitkrai03 please create an issue (and/or PR to fix the infinite issue so they don't run out of stack).

**Comment by jiekang** [comment]
So it looks like the CORS issue still occurs on prod-preview:

```
`Failed to load https://jenkins.prod-preview.openshift.io/job/chrislessard/job/osio-prod-2/job/master/2/?token_json=<token>: Redirect from 'https://jenkins.prod-preview.openshift.io/job/chrislessard/job/osio-prod-2/job/master/2/?token_json=<token>' to 'https://jenkins.prod-preview.openshift.io/job/chrislessard/job/osio-prod-2/job/master/2/' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://prod-preview.openshift.io' is therefore not allowed access.`
```

**Comment by jiekang** [comment]
This means for this to work at all, https://github.com/fabric8-services/fabric8-jenkins-proxy/issues/257 needs to be positively resolved.
